### ISSUE TITLE
Dynaform: Component attribute does not work with datatable

### ISSUE THREAD

**Issue by lttung-axonivy** [primefaces-extensions/primefaces-extensions#799]
**Describe the bug**
Use a component for dynaFormControl of Dynaform, that component has attributes and a datatable. Those attributes resolved to empty values although we pass non-empty values to those attributes.

**Reproducer**
A sample project is at https://github.com/lttung04/dynaform-with-datatable

**Environment:**
 - PF Version: _11.0.0_

**To Reproduce**
1. Go to your deployed application, e.g. http://localhost:8080/primefaces-test/
2. Cannot see a table
![image](https://user-images.githubusercontent.com/76459536/163216786-2f3c9fea-821b-416e-a56c-ec06077889af.png)

**NOTE** Look at the sample code, you could see <h:outputText value="#{cc.attrs.nameColumnRendered}" />, but #{cc.attrs.nameColumnRendered} is **resolved to empty value**.

**Expected behavior**
See a table with one row

**Sample XHTML**
File dynaform.xhtml
```xml
<h:form id="form">
	<local:DocumentContainer />
</h:form>
```
File DocumentContainer.xhtml
```xml
<cc:interface>
</cc:interface>
<cc:implementation>
	<pe:dynaForm id="dyna-form" value="#{dynaFormController.model}" var="dynaData">
		<pe:dynaFormControl type="FileUpload" for="fileUpload">
			<local:DocumentTable id="fileUpload" notInDynaform="true" nameColumnRendered="true" />
		</pe:dynaFormControl>
	</pe:dynaForm>
</cc:implementation>
```
File DocumentTable.xhtml
```xml
<cc:interface>
	<cc:attribute name="nameColumnRendered" default="true" />
	<cc:attribute name="notInDynaform" default="true" />
</cc:interface>
<cc:implementation>
	<cc:renderFacet name="componentHeader" />
	<p:dataTable id="document-table" value="#{dynaFormController.documents}" var="document"
		style="border:solid; border-width:1px">
		<p:column headerText="Name" rendered="#{cc.attrs.nameColumnRendered}">
			<h:outputText id="name" value="#{document}" title="#{document}" />
		</p:column>
	</p:dataTable>
	<h:outputText value="#{cc.attrs.nameColumnRendered}" />
	<cc:renderFacet name="componentFooter" rendered="#{cc.attrs.notInDynaform}" />
</cc:implementation>
```
**Sample Bean**
```java
@ManagedBean
@ViewScoped
public class DynaFormController implements Serializable {

	private static final long serialVersionUID = 1L;
	private DynaFormModel model;
	private List<String> documents;

	public DynaFormController() {
		model = new DynaFormModel();
		documents = new ArrayList<>();
		documents.add("First doc");
		DynaFormRow row = model.createRegularRow();
		row.addControl(row, "FileUpload");
	}

	public DynaFormModel getModel() {
		return model;
	}

	public void setModel(DynaFormModel model) {
		this.model = model;
	}

	public List<String> getDocuments() {
		return documents;
	}

	public void setDocuments(List<String> documents) {
		this.documents = documents;
	}

}

```

**Comment by melloware** [comment]
Did this work on Previous PF versions?  This is inside a composite component as well which adds to the complexity.

**Comment by lttung-axonivy** [comment]
> Did this work on Previous PF versions? This is inside a composite component as well which adds to the complexity.

With PF 7.0.25, primefaces-extension 7.0.3, it works

**Comment by melloware** [comment]
I get diffrent results when running it with `-Pmyfaces23` it looks like it renders something.  Seeing if this is a Mojarra vs MyFaces issue.

**Comment by melloware** [comment]
Can you try removing the Datatable? 

 I am wondering if this is a DynaForm change (as this component has barely changed between 7.0 and 11.0 since no one really maintains the DynaForm) or whether this is actually a Datatable change in core PF.

**Comment by lttung-axonivy** [comment]
If removing Datatable, then it works as expected
```xml
<cc:interface>
	<cc:attribute name="nameColumnRendered" default="true" />
	<cc:attribute name="notInDynaform" default="true" />
</cc:interface>
<cc:implementation>
	<cc:renderFacet name="componentHeader" />
	<h:outputText value="#{cc.attrs.nameColumnRendered}" />
	<cc:renderFacet name="componentFooter" rendered="#{cc.attrs.notInDynaform}" />
</cc:implementation>
```
![image](https://user-images.githubusercontent.com/76459536/163296427-5e133e36-02ba-487e-bf57-d63db4fb0cbe.png)

**Comment by melloware** [comment]
The Datatable went through a massive refactoring between PF 7 and PF 11 and included over 100 bug fixes and improvements.  I have a feeling the issue may be with the Datatable inside your DynaForm inside your CC.  Not so much the DynaForm component itself.

Since this is a complex scenario you may have to debug through the code and figure out where its going wrong.

**Comment by melloware** [comment]
I am going to close this for now as I don't really work on DynaForm and @Rapster was the only one keeping it alive and I think its Datatable related.

**Comment by melloware** [comment]
@lttung04 OK I fixed and stripped down your example:
[pfe-799.zip](https://github.com/primefaces-extensions/primefaces-extensions/files/8640309/pfe-799.zip)

Unzip that and run `mvn clean jetty:run` and navigate to http://localhost:8080/primefaces-test/

It is working fine with a DynaForm with a Datatable.  I stripped out all your composite component stuff.   

In my opinion you have an incredibly messy and complex use case here and you should try and simplify it and remove a lot of that CC stuff.  

This works totally fine:

```xml
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:ui="http://xmlns.jcp.org/jsf/facelets"
      xmlns:f="http://xmlns.jcp.org/jsf/core"
      xmlns:p="http://primefaces.org/ui"
	  xmlns:pe="http://primefaces.org/ui/extensions"
      xmlns:h="http://xmlns.jcp.org/jsf/html">

	<h:head>
		<title>PrimeFaces Test</title>
		<h:outputScript name="test.js" />
	</h:head>
	<h:body>

		<h1>PFE 799</h1>
		<h:form id="frmTest">
			<pe:dynaForm id="dyna-form" value="#{dynaFormController.model}" var="dynaData">
				<pe:dynaFormControl type="FileUpload" for="fileUpload">
					<p:dataTable id="document-table" value="#{dynaFormController.documents}" var="document" style="border:solid; border-width:1px">
						<p:column headerText="Name" rendered="true">
							<h:outputText id="name" value="#{document}" title="#{document}" />
						</p:column>
					</p:dataTable>
				</pe:dynaFormControl>
			</pe:dynaForm>
		</h:form>

	</h:body>
</html>

```

**Comment by lttung-axonivy** [comment]
Hi @melloware 

Thanks for your example, it works as expected but our scenario is quite different. I know this is a quite complex use case. We wanted to reuse component and this is existing system, we migrated from PF7 to PF11 and it is broken.
I mean your sample code works but it is not our use case and migration from PF7 to PF11 for our use case is broken. Could you please analyze and support this use case?

**Comment by melloware** [comment]
@lttung04 OK I dug deeper into this deeper.  Here is a new reproducer where I use your Composite Component inside a DynaFormControl.

Reproducer: [pfe-799-3.zip](https://github.com/primefaces-extensions/primefaces-extensions/files/8661434/pfe-799-3.zip)


```xml
<h1>DyanForm -> Composite -> DataTable</h1>
<h:form id="frmTest">
	<pe:dynaForm id="dyna-form" value="#{dynaFormController.model}" var="dynaData">
		<pe:dynaFormControl type="FileUpload" for="fileUpload">
			<local:DocumentTable id="fileUpload" notInDynaform="true" nameColumnRendered="true" />
		</pe:dynaFormControl>
	</pe:dynaForm>
</h:form>
```

Here is what I found out if you run it.

`mvn clean jetty:run -Pmojarra22` - **FAILS**
`mvn clean jetty:run -Pmojarra23` - **FAILS**
`mvn clean jetty:run -Pmyfaces22` - **WORKS!!!**
`mvn clean jetty:run -Pmyfaces23` - **WORKS!!!**

So basically running 11.0 and PFE 11.0.3 work fine with MyFaces and fails with Mojarra so it is a Mojarra bug.

Could be this issue: https://github.com/eclipse-ee4j/mojarra/issues/5076

Original Issue: https://github.com/primefaces/primefaces/issues/8575

**My opinion is:** don't expect a Mojarra fix and if you can't switch to MyFaces I suggest again making your use case less complicated and removing the Composite Component of a Datatable.

**Comment by lttung-axonivy** [comment]
Hi @melloware 
I have just checked the library, it is myfaces-impl-2.2.13.jar.
Thanks for your sample project. I updated the sample project so that it matches with my scenario Composite > DynaForm -> Composite -> DataTable.
Reproducer: [pfe-799-4.zip](https://github.com/primefaces-extensions/primefaces-extensions/files/8697251/pfe-799-4.zip)

```xml
<h1>Composite > DynaForm -> Composite -> DataTable</h1>
<h:form id="composite-dynaform">
	<local:DocumentContainer />
</h:form>
```
Here is what I found out if you run it.

mvn clean jetty:run -Pmyfaces22 - FAILS with WARNING in console
   `[WARNING] handleException /primefaces-test/ java.io.IOException: parent Composite Component not found when rendering composite component facet composite-dynaform:j_id_9:dyna-form:r1c1p1reg:fileUpload:j_id_i`

mvn clean jetty:run -Pmyfaces23 - WORKS!!!
![Myfaces2 3](https://user-images.githubusercontent.com/76459536/168523363-bfad2173-177c-404b-b3ff-4222ac7a6694.png)

So it means MyFaces2.2 does not work but MyFaces2.3 works.
@melloware Thanks for your analysis and effort. I think we need to consider to use MyFaces2.3 or find a way to solve with MyFaces2.2 like asking for a fix.

**Comment by melloware** [comment]
So yes JSF 2.2 came out in 2010 and JSF 2.3 in 2017.   So it might be worth upgrading to 2.3 since the 2.2 branch is most likely not going to receive much attention especially with JSF 4.0 on the horizon.

**Comment by melloware** [comment]
I meant to say too if you look through Myfaces commits for 2.3 and you figure out which one is the issue I can get it backported to 2.2 branch.

**Comment by melloware** [comment]
OK great news I got to the bottom of it:

1. I kept testing 2.3 versions and discovered it was fixed between **2.3.3** and **2.3.4**.
2. I looked at all commits between those 2 releases and found the one that did it its this one: https://issues.apache.org/jira/browse/MYFACES-4283
3. I then patched a local 2.2.15-SNAPSHOT with that fix and sure enough it fixes your issue!!!

So I am going to backport this fix into **2.2.15**.  If you need 2.2.15 JARs to test let me know.

I just submitted this PR to backport the fix to 2.2.15: https://github.com/apache/myfaces/pull/266

**Comment by melloware** [comment]
[myfaces-impl-2.2.15.zip](https://github.com/primefaces-extensions/primefaces-extensions/files/8747856/myfaces-impl-2.2.15.zip)

Inside that zip is Myfaces 2.2.15 impl JAR if you want to test it out in your environment.

**Comment by lttung-axonivy** [comment]
Hi @melloware 
I tried out myfaces-impl-2.2.15.zip in our environment but it still has error `IOException: parent Composite Component not found when rendering composite component facet form:j_id_k:dyna-form:r1c1p1reg:fileUpload:j_id_t` as we tested before with myfaces-impl-2.2.13.jar.

**Comment by melloware** [comment]
Hmm but it works fine with your reproducer above you provided?  That is what I tested with.

**Comment by melloware** [comment]
I apologize you need the API JAR as well as the Impl.  Here you go...
[myfaces-api-2.2.15.zip](https://github.com/primefaces-extensions/primefaces-extensions/files/8762757/myfaces-api-2.2.15.zip)

**Comment by melloware** [comment]
@lttung04 were you able to verify the API JAR above fixes your issue?

**Comment by lttung-axonivy** [comment]
Hi @melloware 

I have just tested in my environment and it works well with myfaces-impl-2.2.15.zip and myfaces-api-2.2.15.zip. Could you please let me when myfaces-2.2.15 is planned to release? Thanks a lot for your support!

**Comment by melloware** [comment]
That is great to hear!  I posted on the MyFaces users list asking for a release: https://lists.apache.org/list.html?users@myfaces.apache.org

**Comment by melloware** [comment]
@lttung04 MyFaces 2.2.15 is now in Maven Central.  just a warning though this is the last MyFaces 2.2.x release and the branch is being retired so you will eventually have to upgrade to 2.3.

**Comment by lttung-axonivy** [comment]
@melloware Thanks for your information.

**Comment by melloware** [comment]
I have also confirmed this is now fixed in Mojarra 2.3.18 and 4.0.
